components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -438,3 +450,106 @@ func InjectCertConfig(notebook *nbv1.Notebook, configMapName string) error {
 	}
 	return nil
 }
+
+// SetContainerImageFromRegistry checks if there is an internal registry and takes the corresponding actions to set the container.image value.
+// If an internal registry is detected, it uses the default values specified in the Notebook Custom Resource (CR).
+// Otherwise, it checks the last-image-selection annotation to find the image stream and fetches the image from status.dockerImageReference,
+// assigning it to the container.image value.
+func SetContainerImageFromRegistry(ctx context.Context, config *rest.Config, notebook *nbv1.Notebook, log logr.Logger) error {
+    // Create a dynamic client
+    dynamicClient, err := dynamic.NewForConfig(config)
+    if err != nil {
+        log.Error(err, "Error creating dynamic client")
+        return err
+    }
+    // Specify the GroupVersionResource for imagestreams
+    ims := schema.GroupVersionResource{
+        Group:    "image.openshift.io",
+        Version:  "v1",
+        Resource: "imagestreams",
+    }
+
+    annotations := notebook.GetAnnotations()
+    if annotations != nil {
+        if imageSelection, exists := annotations["notebooks.opendatahub.io/last-image-selection"]; exists {
+            // Check if the image selection has an internal registry, if so  will pickup this. This value constructed on the initialization of the Notebook CR.
+            if strings.Contains(notebook.Spec.Template.Spec.Containers[0].Image, "image-registry.openshift-image-registry.svc:5000") {
```

Comment:
Are we sure the container we're interested in is first always in all cases?

Reply:
Another question - are we sure that the value for the internal registry will always be this way? Is there a possibility that there could be some custom value? If so, isn't better to perform check that this value is a non-zero string instead?

Reply:
it is that way the kind: Notebooks with their podspec are assembled, i.e. that the notebook container is always the first one, index 0. But it is not guaranteed.
Here, also, as you can see in my comments elsewhere, the [admission plugin](https://docs.openshift.com/container-platform/4.12/openshift_images/triggering-updates-on-imagestream-changes.html) works with fieldPath-based lookups of the container by container name, which is a good idea.

Reply:
"The other container - oauth-proxy - contains direct quay link from the beginning."
Correct, that is done by odh notebook controller. It is part of an oauth-argument in the odh-notebook-controller-manager pod container.